### ISSUE TITLE
GOG game installs in the wrong place, though it also creates folders in the right place

### ISSUE THREAD

**Issue by OctarineSorcerer** [tkashkin/GameHub#111]
Installed Sunless Skies, and the game installed to `~/GOG` directly, even though my GOG game directory is set to `~/GOG Games/$game`. Two folders were created in `~/GOG Games` in this process. `Sunless Skies`, which contains the installer, and `Sunless_Skies`, which is completely empty.

![~/GOG](https://user-images.githubusercontent.com/6842867/47619295-09acaa00-dad5-11e8-8642-284c53b11328.png)

![image](https://user-images.githubusercontent.com/6842867/47619352-ab33fb80-dad5-11e8-9aab-696c7aa9909a.png)

Same occurred installing Cultist Simulator. Am using the `dev` branch built from source, as of writing

**Comment by OctarineSorcerer** [comment]
Changed GOG Game Directory to `~/GOG/$game`, it still installed directly into `~/GOG` (as opposed to into `~/GOG/Cultist Simulator` or `~/GOG/Cultist_Simulator`

**Comment by tkashkin** [comment]
Make sure you have changed games directory in GOG tab, not in collection tab.
Games are installed in games directory from corresponding settings dialog tab.
Collection directories are used to store downloaded installers and bonus content. Collection directories can have spaces.

**Comment by OctarineSorcerer** [comment]
Ah, okay. Making sure of that change, Cultist Simulator successfully installed to `~/GOG/Cultist_Simulator`. Though, with having both Collection and GOG directory set to the same place, there's behaviour that's a bit inconsistent and confusing - there is a `Cultist Simulator` directory, with the installer contained within, and a `Cultist_Simulator` directory, containing the game.
I can understand that it's due to needing to remove spaces for the game directory, and not for the installer collection, but it does end up confusing. Especially as I imagine that for a game with no spaces, with the same settings, both the installer and game would be placed in the same directory (as opposed to game names with spaces, where they'd be placed in different directories, like Cultist Simulator here)

**Comment by tkashkin** [comment]
Yes, it may be a bit confusing, but I don't think that storing installers and installed games in same directory is a good idea anyway.

It may be possible to install games to some temporary directory and move them after installation to fix this. Or at least some warning should be shown if path contains spaces.

**Comment by OctarineSorcerer** [comment]
Fair point - with the size of installers sometimes being near that of the games themselves, and most of the time only being used once, I wonder if a good place to put them would be in some kind of GameHub cache folder, with an option in the UI to clear it.

As for the temp directory idea, that sounds like a good workaround, as I suspect needing paths without spaces might be a problem for some users.
With the moving after installation, I feel like there should be a warning shown if a folder already resides there, so users don't accidentally nuke something.

**Comment by tkashkin** [comment]
> I wonder if a good place to put them would be in some kind of GameHub cache folder, with an option in the UI to clear it.

Actually it was doing exactly this before. But there are people who want to store their DRM-free game installers and bonus content, maybe on backup drive or NAS.

What I meant to say is that it's a good idea to store installers and installed games in separate directories/partitions/devices and I imagine most of users are doing this. It also means that
> there is a Cultist Simulator directory, with the installer contained within, and a Cultist_Simulator directory, containing the game.

should not be a problem in this case.

**Comment by tkashkin** [comment]
Default paths are here:

https://github.com/tkashkin/GameHub/blob/4150ba217a9a93adfa7f106dea18736d8678272c/data/com.github.tkashkin.gamehub.gschema.xml#L107-L153

I can't think of more appropriate defaults. Storing installers somewhere in `~/.cache` would be more confusing in my opinion.

I'll think about autoremoving option, but how should it work for bonus content?

**Comment by OctarineSorcerer** [comment]
You're right, those are good defaults - sorry for making a fuss about those.

I think it depends on the nature of the bonus content, eg installer for DLC, or data, and that'd be really hard to figure out, so I think it'd be okay to just have autoremoval for game installers, and possibly an indication next to each bonus content, saying if it's installed or not? As is, I can just see a list of bonus content, so I'm not sure if GameHub downloads them all by default and only lists what's installed, or lists all of them anyway.

**Comment by friday** [comment]
`$root/gog-$game` also doesn't work, even though there are no spaces in the path.

> I'll think about autoremoving option, but how should it work for bonus content?

You could call it "Keep installers" or "Remove installer after installing" to make the distinction clear, then make sure the installation didn't fail before removing.

If someone downloads bonus content I think it's fair to assume they want to keep it (unlike pressing "Install" and keeping the installer). In addition, there's no "action" for random content. You can remove installers after installing, but you can't detect when a user has completed consuming a soundtrack or a manual.

**Comment by axredneck** [comment]
I spcified "/home/redneck/games/000_Gamehub" as Collection folder and "/home/redneck/games/000_Gamehub/GOG" as Games folder for GOG but Gamehub still installs games to "/home/redneck/Games"

**Comment by tkashkin** [comment]
@axredneck make sure you have set it [correctly](https://github.com/tkashkin/GameHub/issues/111#issuecomment-433725418): 

> Make sure you have changed games directory in GOG tab, not in collection tab.
Games are installed in games directory from corresponding settings dialog tab.
Collection directories are used to store downloaded installers and bonus content. Collection directories can have spaces.

**Comment by axredneck** [comment]
@tkashkin i already changed directories everywhere.

**Comment by friday** [comment]
Nevermind. Seems to be easy to reproduce (`0.13.0-5e3d6f0-dev`).

This didn't work either:
```
[collection/gog]
bonus='~/Downloads'
installers='/tmp/gamehub-cache'
game-dir='$root/GOG/$game'
dlc='~/Downloads/DLC'

[collection]
root='/home/albin/foo'
```

**Comment by tkashkin** [comment]
@friday I don't see anything wrong here.

What's the output of
`dconf read /com/github/tkashkin/gamehub/paths/gog-games` and
`dconf read /com/github/tkashkin/gamehub/paths/humble-games`?

**Comment by friday** [comment]
I don't have any Humble Bundle PC games.

These are all my paths exported with `dconf dump /com/github/tkashkin/gamehub/paths/`:
```
[/]
gog-games='/home/albin/foo/GOG'
steam-home='/home/albin/.steam'

[collection/gog]
bonus='~/Downloads'
installers='/tmp/gamehub-cache'
game-dir='$root/GOG/$game'
dlc='~/Downloads/DLC'

[collection]
root='/home/albin/foo'
```

I've been testing some more and found that it depends on the games. I think the *games I hadn't installed* and uninstalled before changing the path (including Gemini Rue for Linux, Gobliiins for Linux/Dosbox, Normality for Windows, MDK for Windows and Eschalon: Book I, which I tested for both platforms) worked. One of the games I had installed before (Deponia 2 for Linux) also worked after, but Bloodnet FDD version (Dosbox) and A Bird Story (Linux) and A Golden Wake (Linux) didn't. Bio Menace (Dosbox/Linux) also didn't, but the Windows version did, and the next time the Linux version did too.

Bloodnet FDD version (Dosbox) and A Bird Story (Linux) installs in `~/Games/GOG/$game` regardless of my settings and also regardless of installing the Windows or Linux version. Bloodnet FDD let me change the path if installed with GUI for Windows.
